pkg/healthcheck/healthcheck.go

```diff
@@ -2263,34 +2263,46 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 // Check if there's a pod with the "opaque ports" annotation defined but a
 // service selecting the aforementioned pod doesn't define it
 func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context) error {
+
+	// Pre-fetch services, endpoints and pods for the data plane namespace
+	// to prevent numerous requests to the API server
 	services, err := hc.GetServices(ctx)
 	if err != nil {
 		return err
 	}
 
-	var errStrings []string
+	endpoints, err := hc.GetEndpoints(ctx)
+	if err != nil {
+		return fmt.Errorf("couldn't list endpoints in the cluster: %s", err)
+	}
 
+	dataPlanePods, err := hc.GetDataPlanePods(ctx)
+	if err != nil {
+		return err
+	}
+
+	podsMap := make(map[string]corev1.Pod)
+	for _, pod := range dataPlanePods {
+		podsMap[pod.Name] = pod
+	}
+
+	var errStrings []string
 	for _, service := range services {
 		if service.Spec.ClusterIP == "None" {
 			// skip headless services; they're handled differently
 			continue
 		}
 
-		endpoint, err := hc.kubeAPI.CoreV1().Endpoints(service.Namespace).Get(ctx, service.Name, metav1.GetOptions{})
-
-		if err != nil {
-			return err
-		}
-
+		endpoint := endpoints[service.Namespace+"/"+service.Name]
 		pods := make([]*corev1.Pod, 0)
 		for _, subset := range endpoint.Subsets {
 			for _, addr := range subset.Addresses {
 				if addr.TargetRef != nil && addr.TargetRef.Kind == "Pod" {
-					pod, err := hc.kubeAPI.CoreV1().Pods(service.Namespace).Get(ctx, addr.TargetRef.Name, metav1.GetOptions{})
-					if err != nil {
-						return err
+					pod, ok := podsMap[addr.TargetRef.Name]
+					if !ok {
+						return fmt.Errorf("couldn't find pod for %s", addr.TargetRef.Name)
```

Comment:
This error message looks a bit cryptic to me; we might want to be a bit more explicit about where we couldn't find the pod? This would essentially error out that it can't find a pod for the pod name. Perhaps something that references the endpoints? Just a suggestion tho.

```
return fmt.Errorf("endpoints %s/%s does not contain pod %s", endpoint.Namespace, endpoint.Name, pod.Name)
```

Reply:
Good suggestion for including the endpoints part but shouldn't it be `return fmt.Errorf("could not find pod %s mentioned in the endpoint %s/%s as a target", pod.Name, endpoint.Namespace, endpoint.Name)` as we failed to find the pod instance mentioned in the endpoint? 🤔